### PR TITLE
Type analysis

### PR DESCRIPTION
**Pull request by ajpal**
This PR:
1. Adds type information to eclass analysis data
2. Prevents applying a rule that would cause merging two eclasses with different types
3. Adds typed variables to Pred

Next Up:
1. Add typed variables to String

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -155,9 +158,11 @@ pub trait SynthLanguage: egg::Language + Send + Sync + Display + FromOp + 'stati
         let mut rename_node = |node: &Self| match node.to_var() {
             Some(sym) => {
                 let len = map.len();
-                let var = map
-                    .entry(sym)
-                    .or_insert_with(|| format!("?{}", letter(len)).parse().unwrap());
+                let var = map.entry(sym).or_insert_with(|| {
+                    format!("?{}{}", node.get_type(), letter(len))
```
Comment (chandrakananandi) [inline src/lib.rs:162]: so this makes sure that alpha-renaming preserves the type information, is that right?

**Review by chandrakananandi — APPROVED** [review]
Looks awesome! Thanks for this great PR! I am approving but I did leave some questions and suggestions which would be good to clarify.

File: src/bin/pred.rs [hunk 2]
```diff
@@ -76,23 +152,23 @@ impl SynthLanguage for Pred {
         match self {
             Pred::Lit(c) => vec![Some(c.clone()); cvec_len],
             Pred::Le([x, y]) => {
-                map!(v, x, y => x.to_int().zip(y.to_int()).map(|(x,y)|Constant::Bool(x < y)))
+                map!(v, x, y => Some(Constant::Bool(x.to_int().unwrap() < y.to_int().unwrap())))
```
[the remark this answers is not in the document]
Comment (ajpal) [reply src/bin/pred.rs:155]: yep!

File: src/bin/pred.rs [hunk 3]
```diff
@@ -148,14 +231,18 @@ impl SynthLanguage for Pred {
...
             for _ in 0..10 {
-                vals.push(Some(Constant::Int(rng.gen::<usize>())));
+                i_vals.push(Some(Constant::Int(rng.gen::<usize>())));
+                b_vals.push(Some(Constant::Bool(rng.gen::<bool>())));
             }
-            egraph[id].data.cvec = vals.clone();
+
+            egraph[i_id].data.cvec = i_vals.clone();
+            egraph[b_id].data.cvec = b_vals.clone();
         }
```
[the remark this answers is not in the document]
Comment (ajpal) [reply src/bin/pred.rs:247]: Yes, this is something that's on my to-do list, I just haven't gotten to it yet. If you have cycles and want to do it, that would be great!

File: src/lib.rs (hunk 1 above)
Comment (ajpal) [reply src/lib.rs:162]: Yeah, the variables will be <type-prefix><alpha-renamed symbol>

File: src/bin/pred.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chandrakananandi) [reply src/bin/pred.rs:247]: ok yeah I can take this one!

File: src/bin/pred.rs [hunk 4]
```diff
@@ -66,8 +115,35 @@ impl Constant {
...
+            Type::Int => write!(f, "i"),
+        }
+    }
+}
+
 impl SynthLanguage for Pred {
     type Constant = Constant;
+    type Type = Type;
+
+    fn get_type(&self) -> Self::Type {
```
[the remark this answers is not in the document]
Comment (ajpal) [reply src/bin/pred.rs:137]: I'm not really type-checking here, just annotating the result type of each Op. It's still the responsibility of the user to only supply well-typed terms in the workload

**Pull request merged**